### PR TITLE
expanded /getting-started/writing-documentation.md to include typespecs, added to existing sections

### PR DESCRIPTION
**Pull request by bradhanks**
* tried to "punch up" the intro
* added commonly used elements to the markdown section
* Linked to internal markdown resources. 
* expanded function argument section
* added typespecs section
* added code example to metadata section
* edited and added to recommendation list
* edited and added to Code.fetch_docs/1`section
* added to doctests section

I learned and solidified my understanding about a couple things. Hope some of this helps documentation in the right direction.

### CODE REVIEW

**Comment by josevalim** [comment]
Hi @bradhanks, thank you for the PR. We don't plan to add typespecs to these docs at the moment. This PR is also large and with different type of changes. Can you please submit smaller PRs with focused changes? You can either split per section, per type of change (rewriting vs adding new content) or both. Thanks!

File: lib/elixir/pages/getting-started/writing-documentation.md [hunk 1]
```diff
@@ -1,18 +1,31 @@
...
+  
+  * `#` for H1 headers, `##` for H2 headers, etc.
+  
+  * `*` for unordered list items, or escape numbered list items with a backslash (`1\.`).
 
-Documentation in Elixir is usually attached to module attributes. Let's see an example:
+  * `` ` `` for inline code, or double backticks ` `` ` for inline code with backticks, e.g., ``` `` `elixir` `` ```.
 
-```elixir
+We created an [ExDoc Cheatsheet](https://hexdocs.pm/ex_doc/cheatsheet.html) as a supplemental resource.
```
Comment (josevalim) [inline lib/elixir/pages/getting-started/writing-documentation.md:18]: That's not a markdown cheatsheet. :)

File: lib/elixir/pages/getting-started/writing-documentation.md [hunk 2]
```diff
@@ -1,18 +1,31 @@
-# Writing documentation
-
-Elixir treats documentation as a first-class citizen. Documentation must be easy to write and easy to read. In this guide you will learn how to write documentation in Elixir, covering constructs like module attributes, style practices, and doctests.
+# Writing Documentation
 
+Elixir elevates documentation to first-class status, emphasizing ease of writing as well as readability. This guide will teach you to craft effective documentation in Elixir, covering essential constructs like module attributes, typespecs, style practices, and doctests.
 ## Markdown
 
-Elixir documentation is written using Markdown. There are plenty of guides on Markdown online, we recommend the one from GitHub as a getting started point:
+Elixir documentation is written in Markdown. Some commonly used elements are:
```
Comment (josevalim) [inline lib/elixir/pages/getting-started/writing-documentation.md:6]: We don't plan to add a small markdown summary here. Other resources do it better.

File: lib/elixir/pages/getting-started/writing-documentation.md [hunk 3]
```diff
@@ -1,18 +1,31 @@
-# Writing documentation
-
-Elixir treats documentation as a first-class citizen. Documentation must be easy to write and easy to read. In this guide you will learn how to write documentation in Elixir, covering constructs like module attributes, style practices, and doctests.
+# Writing Documentation
```
Comment (josevalim) [inline lib/elixir/pages/getting-started/writing-documentation.md:1]: We don't use titlecase headings in our docs.

File: lib/elixir/pages/getting-started/writing-documentation.md [hunk 4]
```diff
@@ -1,18 +1,31 @@
...
 
+Elixir elevates documentation to first-class status, emphasizing ease of writing as well as readability. This guide will teach you to craft effective documentation in Elixir, covering essential constructs like module attributes, typespecs, style practices, and doctests.
```
Comment (josevalim) [inline lib/elixir/pages/getting-started/writing-documentation.md:3]: Let's keep the language direct. Converting "write" to "craft" makes it more indirect and imprecise for non-native speakers.

**Comment by bradhanks** [comment]
> Hi @bradhanks, thank you for the PR. We don't plan to add typespecs to these docs at the moment. This PR is also large and with different type of changes. Can you please submit smaller PRs with focused changes? You can either split per section, per type of change (rewriting vs adding new content) or both. Thanks!

Yeah I'm happy to do that. I'm maybe break it out by section I think there are 10 of them.

**Comment by josevalim** [comment]
I have also dropped some comments so you get an idea of which kind of improvements we are looking for. Please avoid rewriting a sentence or paragraph unless there is a clear benefit in doing so (for example, if the current version is unclear or ambiguous or hard to understand). Similar to code, do not rewrite only because of style. Thanks :)

File: lib/elixir/pages/getting-started/writing-documentation.md [hunk 5]
```diff
@@ -49,7 +64,7 @@ def size(%{size: size}) do
 end
 ```
 
-The compiler will infer this argument as `map`. Sometimes the inference will be suboptimal, especially if the function contains multiple clauses with the argument matching on different values each time. You can specify the proper names for documentation by declaring only the function head at any moment before the implementation:
+Sometimes the inference will be suboptimal, especially if the function contains multiple clauses with the argument matching on different values each time. You can specify the proper names for documentation by declaring only the function head at any moment before the implementation.
```
Comment (josevalim) [inline lib/elixir/pages/getting-started/writing-documentation.md:67]: Why did you remove the initial sentence?

Also please avoid small changes such as replacing `:` by `.`. It is very common to use `:` before examples in Elixir. If you change this here, this becomes inconsistent with the rest of the docs.

We need clear and focused changes. Thanks!

File: lib/elixir/pages/getting-started/writing-documentation.md [hunk 6]
```diff
@@ -58,80 +73,156 @@ def size(%{size: size}) do
...
+  * Keep the first paragraph of the documentation a concise description, ideally one line. Tools like [ExDoc](https://github.com/elixir-lang/ex_doc/) use the first line to generate a summary.
 
-    Markdown uses backticks (`` ` ``) to quote code. Elixir builds on top of that to automatically generate links when module or function names are referenced. For this reason, always use full module names. If you have a module called `MyApp.Hello`, always reference it as `` `MyApp.Hello` `` and never as `` `Hello` ``.
+  * Reference modules by their full name and wrap them in backticks (`` ` ``). Elixir builds on top of Markdown to automatically generate links for modules and functions, so reference `MyApp.Hello` as `` `MyApp.Hello` ``, never as `` `Hello` ``.
+  
+  * Reference local functions by their name and arity, as in `` `world/1` ``.
 
-  * Reference functions by name and arity if they are local, as in `` `world/1` ``, or by module, name and arity if pointing to an external module: `` `MyApp.Hello.world/1` ``.
+  * Reference external functions by their module name followed by the function name and arity, e.g., `` `MyApp.Hello.world/1` ``.
```
Comment (josevalim) [inline lib/elixir/pages/getting-started/writing-documentation.md:154]: We don't use `e.g.` in docs, for clarity. :)

File: lib/elixir/pages/getting-started/writing-documentation.md [hunk 7]
```diff
@@ -58,80 +73,156 @@ def size(%{size: size}) do
...
+```
 
-When writing documentation:
+## Recommendations
 
-  * Keep the first paragraph of the documentation concise and simple, typically one-line. Tools like [ExDoc](https://github.com/elixir-lang/ex_doc/) use the first line to generate a summary.
+When writing documentation, we recommend you follow these guidelines:
 
-  * Reference modules by their full name.
+  * Keep the first paragraph of the documentation a concise description, ideally one line. Tools like [ExDoc](https://github.com/elixir-lang/ex_doc/) use the first line to generate a summary.
```
Comment (josevalim) [inline lib/elixir/pages/getting-started/writing-documentation.md:148]: This is pretty much the same as before, but written differently. Please avoid such changes.

**Comment by josevalim** [comment]
I pushed a couple more comments.

> Yeah I'm happy to do that. I'm maybe break it out by section I think there are 10 of them.

Based on the notes so far, I don't think you will need 10 pull requests. :) Most changes will be rejected unless they are arguably improving the docs, so I recommend focusing on the important ones.

It is important to avoid churn, otherwise we would end-up rewriting the docs quite frequently!

**Comment by bradhanks** [comment]
> I pushed a couple more comments.
> 
> 
> 
> > Yeah I'm happy to do that. I'm maybe break it out by section I think there are 10 of them.
> 
> 
> 
> Based on the notes so far, I don't think you will need 10 pull requests. :) Most changes will be rejected unless they are arguably improving the docs, so I recommend focusing on the important ones.
> 
> 
> 
> It is important to avoid churn, otherwise we would end-up rewriting the docs quite frequently!

Haha that makes sense based on what you are saying. I'm not planning on 10 just theoretically they would probably be better at 10% the size 😀
